### PR TITLE
Improve configurability with new rpc messages

### PR DESCRIPTION
**Pull request by smolck**
This mostly works, with only one current problem: when the user disables the ext_tabline while it is in use, it doesn't disappear (and couldn't be made to go away through closing tabs when I tested), so that he ends up with something like this (gnvim's tabs won't go away):

image

Also, after disabling the ext_tabline the TUI tabline doesn't appear until the user changes tabs (at least when I tested it). I can work on fixing this once we know how we want to go about doing so (I don't want to bloat the code base with a bad solution). (As an FYI I am using Neovim master from about a week ago, give or take.) I don't think this could be a problem with the popupmenu, since one can't call the commands to enable/disable it without exiting insert mode (may be wrong though).

If we want to add any other configuration options, let me know and I'll add them. Any feedback is much appreciated. (Note that once things are solidified here, we’ll probably want to add documentation for the new commands. Don’t want to do that just yet though, since we may want to change the command names, add more commands, etc.)

See #59.

### CODE REVIEW

**Comment by smolck** [comment]
So, this diff mostly fixes the problem with the tabline not going away when it is disabled (it is at the bottom of the `GnvimEvent::DisableExtTabline` match arm):
```diff
diff --git a/src/ui/ui.rs b/src/ui/ui.rs
index 9c76c5b..fc31586 100644
--- a/src/ui/ui.rs
+++ b/src/ui/ui.rs
@@ -445,6 +445,7 @@ fn handle_gnvim_event(
                     ));
                 },
             );
+            state.tabline.get_widget().hide();
         }
         GnvimEvent::DisableExtPopupmenu => {
             let mut nvim = nvim.lock().unwrap();
```

Only problem left is that if buffers are shown in the TUI tabline, which is possible with vim-airline, the area where the TUI tabline previously was shown is blank until the window is updated (i.e. by changing tabs, creating a split, etc., even resizing the window I think) when the external tabline is re-enabled.

**Comment by badosu** [comment]
Perhaps sending a `RedrawEvent::TablineUpdate` fixes this?

**Comment by smolck** [comment]
How should I send that `RedrawEvent`? I can't really call `handle_redraw_event` directly with the current setup (and if I could I'd need the tabpages and the strings associated with them).

**Comment by badosu** [comment]
Hmm, I thought exactly about that method, we should at least have this available for these cases so might be necessary to build it if not accessible now.

**Comment by smolck** [comment]
@badosu That should work; I’ll give it a try. There’s probably API methods for getting the info we want, so this may be pretty simple. I’ll post a diff if I get one working.

**Comment by smolck** [comment]
The problem I'm now running into is getting the name of each tab. That information is only gotten from [here](https://github.com/vhakulinen/gnvim/blob/master/src/nvim_bridge/mod.rs#L811) AFAIK. Of course, I could always find a way to pass that information around, but that feels unnecessary. I can get the tabpages and the current tabpage via API methods, but the name of the Tabpage is not easily accessible ATM. I may be missing something though; any ideas?

**Comment by kenardes** [comment]
hello, noobs here, how to show / enable vim-airline tabline? it's shown on TUI / terminal, but not in gnvim. I need it to easily see buffer list.

**Comment by smolck** [comment]
@kenardes Until this PR is merged, try running GNvim like this:

`$ gnvim --disable-ext-tabline`

If that doesn't work or I misunderstood, please let me know.

**Comment by kenardes** [comment]
> If that doesn't work or I misunderstood, please let me know.

it work, thanks.
but is there away to apply same option "disable-ext-tabline" in vim config file ?

**Comment by smolck** [comment]
> but is there away to apply same option "disable-ext-tabline" in vim config file ?

At the moment, master has no way of doing that. However, if you're willing to use this branch (note that it isn't done yet), then you can put this in your `.vimrc`/`init.vim`:

```vim
if exists('g:gnvim')
    call gnvim#tabline#disable()
endif
```

But again, the above will only work with this branch.

@vhakulinen Do you mind giving this branch a review once you get the chance? I know that documentation needs to be taken care of, but is there anything else?

**Comment by vhakulinen** [comment]
Only note so far: remove the `Disable...` messages, and make the `Enable...` messages to take a boolean argument which indicates if given feature should be enabled or not. The cli arguments can stay (for now at least).

P.S. I'm going to start my holidays in a week, so hopefully I find the energy to invest my time into gnvim again.

**Comment by smolck** [comment]
@vhakulinen @badosu Besides the conflicts and maybe some other small things, the main problem with this branch right now is the problem with the tabline mentioned above. @badosu suggested manually issuing a redraw event for the tabline; is that the direction we want to go? If so, we’ll need a way to get the necessary info for issuing the redraw event (getting that info’s where I got stuck).

**Comment by smolck** [comment]
This branch is now ready for testing and review. I fixed the problems with the tabline, using @badosu's suggestion of getting the information for the tabline and manually updating it. Let me know if you spot any issues, inconsistencies, etc.

cc: @vhakulinen

File: src/nvim_bridge/mod.rs [hunk 1]
```diff
@@ -820,6 +820,10 @@ impl fmt::Display for RedrawEvent {
 #[derive(Debug, PartialEq)]
 pub enum GnvimEvent {
     SetGuiColors(SetGuiColors),
+    EnableExtTabline(bool),
+    EnableExtCmdline(bool),
+    EnableExtPmenu(bool),
```
Comment (smolck) [inline src/nvim_bridge/mod.rs]: Should these say `DisplayExt ...` instead of `EnableExt ...`?

File: src/ui/ui.rs [hunk 2]
```diff
@@ -408,6 +408,66 @@ fn handle_gnvim_event(
...
+                        err
+                    ));
+                });
+
+            if !*option {
+                state.tabline.get_widget().hide();
+                return;
+            }
+
+            match (nvim.get_current_tabpage(), nvim.list_tabpages()) {
```
Comment (smolck) [inline src/ui/ui.rs]: This error handling may be excessive, if so let me know.

File: src/ui/ui.rs [hunk 3]
```diff
@@ -408,6 +408,66 @@ fn handle_gnvim_event(
...
+                    let tabpages = tabpages
+                        .iter()
+                        .map(|x| (x.clone(), String::from("")))
+                        .collect();
+
+                    state.tabline.update(&mut nvim, current, tabpages);
+                }
+                (Err(err), Ok(_)) => {
+                    nvim.command_async(&format!(
+                            "echo \"Unable to get current tabpage for updating tabline: '{}'\"", err));
```
Comment (smolck) [inline src/ui/ui.rs]: Better error msg?

File: src/ui/ui.rs [hunk 4]
```diff
@@ -408,6 +408,66 @@ fn handle_gnvim_event(
...
+                }
+                (Ok(_), Err(err)) => {
+                    nvim.command_async(&format!(
+                            "echo \"Unable to get tabpages for updating tabline: '{}'\"", err));
```
Comment (smolck) [inline src/ui/ui.rs]: Same as above.

File: src/ui/ui.rs [hunk 5]
```diff
@@ -408,6 +408,66 @@ fn handle_gnvim_event(
...
+                }
+                (Err(get_current_err), Err(list_tabpages_err)) => {
+                    nvim.command_async(&format!(
+                            "echo \"Unable to get current tabpage for updating tabline: '{}'\n
+                            Unable to get tabpages for updating tabline: '{}'\"", get_current_err, list_tabpages_err));
```
Comment (smolck) [inline src/ui/ui.rs]: Same as above.

**Comment by Famiu** [comment]
Why is this not merged yet? Currently it's a pain to disable GNvim features, I feel like this should be merged asap.

**Pull request closed without merging**